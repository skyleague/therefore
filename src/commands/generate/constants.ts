import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
