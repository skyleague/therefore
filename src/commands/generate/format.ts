import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import { memoize } from '@skyleague/axioms'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

export async function formatContent({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    return input
}

const execAsync = promisify(exec)

const getBiomeBinaryPath = memoize(async () => {
    try {
        const { stdout } = await execAsync('npx --no-install biome --version')
        if (stdout.trim()) {
            return 'npx'
        }
    } catch (error) {
        console.error('Error finding Biome:', error)
    }
    return null
})

export async function formatBiomeFiles(files: string[]) {
    const biomePath = await getBiomeBinaryPath()
    if (!biomePath) {
        console.warn('Biome not found. Skipping formatting.')
        return
    }

    const baseCommand = `${biomePath} biome check --write `
    let fileChunks: string[][]
    if (process.platform === 'win32') {
        // Windows command line length limit
        const MAX_COMMAND_LENGTH = 8180
        const remainingLength = MAX_COMMAND_LENGTH - baseCommand.length

        fileChunks = []
        let currentChunk: string[] = []
        let currentLength = 0

        for (const file of files) {
            // +1 for the space between files
            // +2 for the quotes
            const fileLength = file.length + 1 + 2

            if (currentLength + fileLength > remainingLength) {
                fileChunks.push(currentChunk)
                currentChunk = []
                currentLength = 0
            }

            currentChunk.push(`"${file}"`)
            currentLength += fileLength
        }

        if (currentChunk.length > 0) {
            fileChunks.push(currentChunk)
        }
    } else {
        // On non-Windows platforms, process all files at once
        fileChunks = [files]
    }

    try {
        const promises = fileChunks.map(async (fileChunk) => {
            const command = `${baseCommand} ${fileChunk.join(' ')}`
            const { stdout, stderr } = await execAsync(command)

            if (stderr) {
                console.warn('Biome formatting warning:', stderr)
            }
            return stdout
        })

        const results = await Promise.all(promises)
        console.log('Biome formatting completed:', results.join('\n'))
    } catch (error) {
        console.error('Error running Biome format:', error)
    }
}
