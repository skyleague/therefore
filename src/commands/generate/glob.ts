import { Nothing, fstat, isJust, isRight, normalizePath } from '@skyleague/axioms'
import fastGlob from 'fast-glob'

import path from 'node:path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.allSettled(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && isJust(eitherStat.right)) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(normalizePath(pattern))
                    }
                    if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(normalizePath(path.relative(cwd, absolutePath)))}/**/*${extension}`
                    }

                    return Nothing
                }
                if (pattern.startsWith('!')) {
                    globOptions.ignore.push(normalizePath(pattern.slice(1)))
                    return Nothing
                }
                return normalizePath(pattern)
            }),
        )
    )
        .map((result) => (result.status === 'fulfilled' ? result.value : Nothing))
        .filter(isJust)
    return await fastGlob(entries, globOptions)
}
