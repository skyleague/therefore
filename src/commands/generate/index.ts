import { generate } from './generate'

import { getExtension } from '../../common/template/path'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('compile', {
            default: true,
            type: 'boolean',
        })
        .option('ext', {
            default: '.schema.ts',
            type: 'string',
        })
        .option('out-ext', {
            default: '.type.ts',
            type: 'string',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, ext, 'out-ext': outExt, compile, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((i) => i.toString()),
        extension: ext,
        outputFileRename: (path: string) => path.replace(getExtension(path)!, outExt),
        compile,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
