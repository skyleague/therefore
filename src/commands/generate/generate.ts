import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
