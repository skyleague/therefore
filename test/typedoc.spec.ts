import { compileOutputFiles } from '../src/commands/generate/generate'

describe('schema', () => {
    test('typedoc', async () => {
        expect(
            await compileOutputFiles(['examples/typedoc/typedoc.schema.ts'], {
                outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
                cwd: process.cwd(),
                compile: true,
            })
        ).toMatchSnapshot()
    })
})
