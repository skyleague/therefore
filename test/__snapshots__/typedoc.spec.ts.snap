// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema typedoc 1`] = `
Array [
  Object {
    "data": Object {
      "00010-000": Object {
        "referenceName": "{{00010-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Src",
        "uniqueSymbolName": "{{00010-000:symbolName}}7ac8",
      },
      "00013-000": Object {
        "referenceName": "{{00013-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Theme",
        "uniqueSymbolName": "{{00013-000:symbolName}}7ac8",
      },
      "0003-000": Object {
        "referenceName": "{{0003-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Plugin",
        "uniqueSymbolName": "{{0003-000:symbolName}}7ac8",
      },
      "00052-000": Object {
        "referenceName": "{{00052-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Typedoc",
        "uniqueSymbolName": "{{00052-000:symbolName}}",
      },
      "0006-000": Object {
        "referenceName": "{{0006-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Readme",
        "uniqueSymbolName": "{{0006-000:symbolName}}7ac8",
      },
    },
    "prettify": true,
    "relativeSource": "./typedoc.schema",
    "targetPath": "examples/typedoc/typedoc.schema.type.ts",
    "template": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */
import AjvValidator from 'ajv'
import type { ValidateFunction } from 'ajv'


/**
 * JSON Schema for typedoc.json
 */
export interface {{00052-000:symbolName}} {
    /**
     * Should TypeDoc disable the testing and cleaning of the output directory?
     * 
     * @default false
     */
    disableOutputCheck: boolean
    /**
     * Specifies the fully qualified name of the root symbol.
     * 
     * @default './'
     */
    entryPoint: string
    /**
     * Exclude files by the given pattern when a path is provided as source. Supports minimatch patterns.
     */
    exclude: (string)[]
    /**
     * Prevent externally resolved TypeScript files from being documented.
     * 
     * @default false
     */
    excludeExternals: boolean
    /**
     * Prevent symbols that are not exported from being documented.
     * 
     * @default false
     */
    excludeNotExported: boolean
    /**
     * Ignores private variables and methods
     * 
     * @default false
     */
    excludePrivate: boolean
    /**
     * Ignores protected variables and methods
     * 
     * @default false
     */
    excludeProtected: boolean
    /**
     * Define a pattern for files that should be considered being external.
     */
    externalPattern: (string)[]
    /**
     * Set the Google Analytics tracking ID and activate tracking code.
     */
    gaID: string
    /**
     * Set the site name for Google Analytics.
     * 
     * @default 'auto'
     */
    gaSite: string
    /**
     * Use specified revision or branch instead of the last revision for linking to GitHub source files.
     */
    gitRevision: string
    /**
     * Do not print the TypeDoc link at the end of the page.
     * 
     * @default false
     */
    hideGenerator: boolean
    /**
     * Generates documentation, even if the project does not TypeScript compile.
     * 
     * @default false
     */
    ignoreCompilerErrors: boolean
    /**
     * Turn on parsing of .d.ts declaration files.
     * 
     * @default false
     */
    includeDeclarations: boolean
    /**
     * Specifies the location to look for included documents (use [[include:FILENAME]] in comments).
     */
    includes: string
    /**
     * The sources files from which to build documentation.
     */
    inputFiles: (string)[]
    /**
     * Specifies the location to output a JSON file containing all of the reflection data.
     */
    json: string
    /**
     * Emits a list of broken symbol [[navigation]] links after documentation generation
     * 
     * @default false
     */
    listInvalidSymbolLinks: boolean
    /**
     * Specify the logger that should be used.
     * 
     * @default 'console'
     */
    logger: 'console' | 'none'
    /**
     * Specifies the location with media files that should be copied to the output directory.
     */
    media: string
    /**
     * Specifies the output mode the project is used to be compiled with.
     * 
     * @default 'modules'
     */
    mode: 'file' | 'modules'
    /**
     * Set the name of the project that will be used in the header of the template.
     */
    name: string
    /**
     * Specifies the location the documentation should be written to.
     */
    out: string
    plugin: {{0003-000:referenceName}}
    /**
     * Add the package version according to package.json to the projects name.
     * 
     * @default false
     */
    includeVersion: boolean
    /**
     * Specify tags that should be removed from doc comments when parsing.
     * 
     * @default []
     */
    excludeTags: (string)[]
    readme: {{0006-000:referenceName}}
    src: {{00010-000:referenceName}}
    /**
     * Remove reflections annotated with @internal
     * 
     * @default false
     */
    stripInternal: boolean
    theme: {{00013-000:referenceName}}
    /**
     * Specifies the top level table of contents.
     */
    toc: (string)[]
    /**
     * Specify a typescript config file that should be loaded. If not specified TypeDoc will look for 'tsconfig.json' in the current directory.
     * 
     * @default './tsconfig.json'
     */
    tsconfig: string
}

export const {{00052-000:symbolName}} = {
    validate: require('./schemas/typedoc.schema.js') as ValidateFunction<{{00052-000:symbolName}}>,
    get schema() { return {{00052-000:symbolName}}.validate.schema},
    source: \`\${__dirname}/typedoc.schema\`,
    sourceSymbol: 'typedoc',
    is: (o: unknown): o is {{00052-000:symbolName}} => {{00052-000:symbolName}}.validate(o) === true,
    assert: (o: unknown) => {
        if (!{{00052-000:symbolName}}.validate(o)) {
            throw new AjvValidator.ValidationError({{00052-000:symbolName}}.validate.errors ?? [])
        }
    },
} as const

/**
 * Specify the npm plugins that should be loaded. Omit to load all installed plugins.
 */
type {{0003-000:symbolName}} = (string)[]

/**
 * Path to the readme file that should be displayed on the index page. Pass none to disable the index page and start the documentation on the globals page.
 */
type {{0006-000:symbolName}} = 'none' | string

/**
 * The sources files from which to build documentation.
 * DEPRECATED: Use inputFiles instead.
 */
type {{00010-000:symbolName}} = string | (string)[]

/**
 * Specify the path to the theme that should be used.
 */
type {{00013-000:symbolName}} = 'default' | 'minimal' | string

",
    "type": "typescript",
  },
  Object {
    "data": Object {
      "00010-000": Object {
        "referenceName": "{{00010-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Src",
        "uniqueSymbolName": "{{00010-000:symbolName}}7ac8",
      },
      "00013-000": Object {
        "referenceName": "{{00013-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Theme",
        "uniqueSymbolName": "{{00013-000:symbolName}}7ac8",
      },
      "0003-000": Object {
        "referenceName": "{{0003-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Plugin",
        "uniqueSymbolName": "{{0003-000:symbolName}}7ac8",
      },
      "00052-000": Object {
        "referenceName": "{{00052-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Typedoc",
        "uniqueSymbolName": "{{00052-000:symbolName}}",
      },
      "0006-000": Object {
        "referenceName": "{{0006-000:symbolName}}",
        "srcPath": "examples/typedoc/typedoc.schema.type.ts",
        "symbolName": "Readme",
        "uniqueSymbolName": "{{0006-000:symbolName}}7ac8",
      },
    },
    "prettify": true,
    "relativeSource": "./typedoc.schema",
    "targetPath": "examples/typedoc/schemas/typedoc.schema.json",
    "template": "{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"type\\": \\"object\\",
  \\"title\\": \\"JSON Schema for typedoc.json\\",
  \\"properties\\": {
    \\"disableOutputCheck\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Should TypeDoc disable the testing and cleaning of the output directory?\\",
      \\"default\\": false
    },
    \\"entryPoint\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the fully qualified name of the root symbol.\\",
      \\"default\\": \\"./\\"
    },
    \\"exclude\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Exclude files by the given pattern when a path is provided as source. Supports minimatch patterns.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"excludeExternals\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Prevent externally resolved TypeScript files from being documented.\\",
      \\"default\\": false
    },
    \\"excludeNotExported\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Prevent symbols that are not exported from being documented.\\",
      \\"default\\": false
    },
    \\"excludePrivate\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Ignores private variables and methods\\",
      \\"default\\": false
    },
    \\"excludeProtected\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Ignores protected variables and methods\\",
      \\"default\\": false
    },
    \\"externalPattern\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Define a pattern for files that should be considered being external.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"gaID\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the Google Analytics tracking ID and activate tracking code.\\"
    },
    \\"gaSite\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the site name for Google Analytics.\\",
      \\"default\\": \\"auto\\"
    },
    \\"gitRevision\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Use specified revision or branch instead of the last revision for linking to GitHub source files.\\"
    },
    \\"hideGenerator\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Do not print the TypeDoc link at the end of the page.\\",
      \\"default\\": false
    },
    \\"ignoreCompilerErrors\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Generates documentation, even if the project does not TypeScript compile.\\",
      \\"default\\": false
    },
    \\"includeDeclarations\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Turn on parsing of .d.ts declaration files.\\",
      \\"default\\": false
    },
    \\"includes\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location to look for included documents (use [[include:FILENAME]] in comments).\\"
    },
    \\"inputFiles\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"The sources files from which to build documentation.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"json\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location to output a JSON file containing all of the reflection data.\\"
    },
    \\"listInvalidSymbolLinks\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Emits a list of broken symbol [[navigation]] links after documentation generation\\",
      \\"default\\": false
    },
    \\"logger\\": {
      \\"description\\": \\"Specify the logger that should be used.\\",
      \\"default\\": \\"console\\",
      \\"enum\\": [
        \\"console\\",
        \\"none\\"
      ]
    },
    \\"media\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location with media files that should be copied to the output directory.\\"
    },
    \\"mode\\": {
      \\"description\\": \\"Specifies the output mode the project is used to be compiled with.\\",
      \\"default\\": \\"modules\\",
      \\"enum\\": [
        \\"file\\",
        \\"modules\\"
      ]
    },
    \\"name\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Set the name of the project that will be used in the header of the template.\\"
    },
    \\"out\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specifies the location the documentation should be written to.\\"
    },
    \\"plugin\\": {
      \\"$ref\\": \\"#/$defs/{{0003-000:uniqueSymbolName}}\\"
    },
    \\"includeVersion\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Add the package version according to package.json to the projects name.\\",
      \\"default\\": false
    },
    \\"excludeTags\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Specify tags that should be removed from doc comments when parsing.\\",
      \\"default\\": [],
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"readme\\": {
      \\"$ref\\": \\"#/$defs/{{0006-000:uniqueSymbolName}}\\"
    },
    \\"src\\": {
      \\"$ref\\": \\"#/$defs/{{00010-000:uniqueSymbolName}}\\"
    },
    \\"stripInternal\\": {
      \\"type\\": \\"boolean\\",
      \\"description\\": \\"Remove reflections annotated with @internal\\",
      \\"default\\": false
    },
    \\"theme\\": {
      \\"$ref\\": \\"#/$defs/{{00013-000:uniqueSymbolName}}\\"
    },
    \\"toc\\": {
      \\"type\\": \\"array\\",
      \\"description\\": \\"Specifies the top level table of contents.\\",
      \\"items\\": {
        \\"type\\": \\"string\\"
      }
    },
    \\"tsconfig\\": {
      \\"type\\": \\"string\\",
      \\"description\\": \\"Specify a typescript config file that should be loaded. If not specified TypeDoc will look for 'tsconfig.json' in the current directory.\\",
      \\"default\\": \\"./tsconfig.json\\"
    }
  },
  \\"required\\": [
    \\"disableOutputCheck\\",
    \\"entryPoint\\",
    \\"exclude\\",
    \\"excludeExternals\\",
    \\"excludeNotExported\\",
    \\"excludePrivate\\",
    \\"excludeProtected\\",
    \\"externalPattern\\",
    \\"gaID\\",
    \\"gaSite\\",
    \\"gitRevision\\",
    \\"hideGenerator\\",
    \\"ignoreCompilerErrors\\",
    \\"includeDeclarations\\",
    \\"includes\\",
    \\"inputFiles\\",
    \\"json\\",
    \\"listInvalidSymbolLinks\\",
    \\"logger\\",
    \\"media\\",
    \\"mode\\",
    \\"name\\",
    \\"out\\",
    \\"plugin\\",
    \\"includeVersion\\",
    \\"excludeTags\\",
    \\"readme\\",
    \\"src\\",
    \\"stripInternal\\",
    \\"theme\\",
    \\"toc\\",
    \\"tsconfig\\"
  ],
  \\"additionalProperties\\": false,
  \\"$defs\\": {
    \\"{{0003-000:uniqueSymbolName}}\\": {
      \\"title\\": \\"plugin\\",
      \\"description\\": \\"Specify the npm plugins that should be loaded. Omit to load all installed plugins.\\",
      \\"oneOf\\": [
        {
          \\"type\\": \\"array\\",
          \\"default\\": [
            \\"none\\"
          ],
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      ]
    },
    \\"{{0006-000:uniqueSymbolName}}\\": {
      \\"title\\": \\"readme\\",
      \\"description\\": \\"Path to the readme file that should be displayed on the index page. Pass none to disable the index page and start the documentation on the globals page.\\",
      \\"oneOf\\": [
        {
          \\"const\\": \\"none\\"
        },
        {
          \\"type\\": \\"string\\"
        }
      ]
    },
    \\"{{00010-000:uniqueSymbolName}}\\": {
      \\"title\\": \\"src\\",
      \\"description\\": \\"The sources files from which to build documentation.\\\\nDEPRECATED: Use inputFiles instead.\\",
      \\"oneOf\\": [
        {
          \\"type\\": \\"string\\"
        },
        {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"type\\": \\"string\\"
          }
        }
      ]
    },
    \\"{{00013-000:uniqueSymbolName}}\\": {
      \\"title\\": \\"theme\\",
      \\"description\\": \\"Specify the path to the theme that should be used.\\",
      \\"oneOf\\": [
        {
          \\"enum\\": [
            \\"default\\",
            \\"minimal\\"
          ]
        },
        {
          \\"type\\": \\"string\\"
        }
      ]
    }
  }
}",
    "type": "jsonschema",
  },
]
`;
